pkg/logentry/stages/multiline.go

```diff
@@ -0,0 +1,195 @@
+)
+
+const maxLineDefault uint64 = 128
+
+// MultilineConfig contains the configuration for a multilineStage
+type MultilineConfig struct {
+	Expression  *string `mapstructure:"firstline"`
+	regex       *regexp.Regexp
+	MaxLines    *uint64 `mapstructure:"max_lines"`
+	MaxWaitTime *string `mapstructure:"max_wait_time"`
+	maxWait     time.Duration
+}
+
+func validateMultilineConfig(cfg *MultilineConfig) error {
+	if cfg == nil || cfg.Expression == nil || cfg.MaxWaitTime == nil {
+		return errors.New(ErrMultilineStageEmptyConfig)
+	}
+
+	expr, err := regexp.Compile(*cfg.Expression)
+	if err != nil {
+		return errors.Errorf(ErrMultilineStageInvalidRegex, err)
+	}
+	cfg.regex = expr
+
+	maxWait, err := time.ParseDuration(*cfg.MaxWaitTime)
+	if err != nil {
+		return errors.Errorf(ErrMultilineStageInvalidMaxWaitTime, err)
+	}
+	cfg.maxWait = maxWait
+
+	if cfg.MaxLines == nil {
+		cfg.MaxLines = new(uint64)
+		*cfg.MaxLines = maxLineDefault
+	}
+
+	return nil
+}
+
+// dropMultiline matches lines to determine whether the following lines belong to a block and should be collapsed
+type multilineStage struct {
+	logger log.Logger
+	cfg    *MultilineConfig
+}
+
+// multilineState captures the internal state of a running multiline stage.
+type multilineState struct {
+	buffer         *bytes.Buffer // The lines of the current multiline block.
+	startLineEntry Entry         // The entry of the start line of a multiline block.
+	currentLines   uint64        // The number of lines of the current multiline block.
+}
+
+// newMulitlineStage creates a MulitlineStage from config
+func newMultilineStage(logger log.Logger, config interface{}) (Stage, error) {
+	cfg := &MultilineConfig{}
+	err := mapstructure.WeakDecode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	err = validateMultilineConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &multilineStage{
+		logger: log.With(logger, "component", "stage", "type", "multiline"),
+		cfg:    cfg,
+	}, nil
+}
+
+func (m *multilineStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+
+		streams := make(map[model.Fingerprint](chan Entry))
+		wg := new(sync.WaitGroup)
+
+		for e := range in {
+			key := e.Labels.FastFingerprint()
+			s, ok := streams[key]
+			if !ok {
+				level.Debug(m.logger).Log("msg", "creating new stream", "stream", key)
+				s = make(chan Entry)
+				streams[key] = s
+
+				wg.Add(1)
+				go m.runMultiline(s, out, wg)
+			}
+			level.Debug(m.logger).Log("msg", "pass entry", "stream", key, "line", e.Line)
+			s <- e
+		}
+
+		// Close all streams and wait for them to finish being processed.
+		for _, s := range streams {
+			close(s)
+		}
+		wg.Wait()
+	}()
+	return out
+}
+
+func (m *multilineStage) runMultiline(in chan Entry, out chan Entry, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	state := &multilineState{
+		buffer:       new(bytes.Buffer),
+		currentLines: 0,
+	}
+
+	for {
+		select {
+		case <-time.After(m.cfg.maxWait):
```

Comment:
I think we should also exit the goroutine after some time, a stream could stop sending data because the upstream pod/file/application is dead and a new one could start, over a long period of time we could end up with a lot of staled goroutines.

This could be after x cycle of maxWait that didn't flush anything. Careful you'll need to protect that map with a lock now and you might have a race when closing all stream vs one stream closing itself.

Reply:
> because the upstream pod/file/application is dead and a new one could start

Shouldn't the input channel close in this case? Ideally we would have a timer for the whole pipeline. What do you think?

Reply:
In most cases the `in` would close, for the file case for instance yep. I'm not sure what would happen for syslog or other types of targets.

We can go like this and add an issue to cleanup goroutine when a stream go stale, and prioritize if someone has this issue later.

Another way to see it is should we make the goroutine only if `isFirstLine := m.cfg.regex.MatchString(e.Line)` is true ?

Reply:
That's a good point. Right now we would wait some time and try to flush an empty block and wait again. We only create a new goroutine when a new line with a new finger print arrives. I think we do need the first goroutine to manage the incoming lines and labels though. Should we add a timeout there? I.e. if nothing arrives we close? Which we might not want to do either though since the stage above might be sending new entries.

Reply:
Hm, I'd say we land this and create an issue to cleanup goroutines. It feels this is more evolved.

Reply:
Let's do that for the cleanup but still check my other comment.